### ISSUE TITLE
https proxy call brakes logo.svg URL (also some strange redirect removing /piwik/ folder) and getting a "No input file specified." response.

### ISSUE THREAD

**Issue by courtens** [piwik/piwik#11765]
There are three issues I have, and they are all related to the proxy setup problem I am experiencing.

1) ======
On the login page located here `https://www.mydomainname.com/piwik/`  I get a broken logo image. The logo image is located on the server at `http://192.168.1.1:8080/plugins/Morpheus/images/logo.svg`. But somehow the proxy (or piwik) responds that the logo.svg file is located here: `https://www.mydomainname.com/plugins/Morpheus/images/logo.svg` which is wrong, where in fact it should be `https://www.mydomainname.com/piwik/plugins/Morpheus/images/logo.svg`. Somehow `/piwik/` gets dropped.

2) ======
Immediately after logging in I get a redirect to a page that prompts: `No input file specified.` (or a redirect to  `https://www.mydomainname.com/` instead of  `https://www.mydomainname.com/piwik/`) It is only after clicking back on the browser that I get redirected to the piwik website.  Somehow the piwik code truncates the piwik folder during the redirect right after a successful login (back to the root folder.) This redirect brakes the page.

3) =====
The "All Websites" page is not working (not giving me any data), only the "Dashboard" page works. "Add a new website" goes to `https://www.mydomainname.com/?module=SitesManager&action=index&showaddsite=1&period=range&date=previous30&idSite=1` but it should go to  `https://www.mydomainname.com/piwik/?module=SitesManager&action=index&showaddsite=1&period=range&date=previous30&idSite=1`

Same issue as issue 1 and 2, a redirect dropping the working folder I am in.

=== my setup ============
I am running Ubuntu, hosting VM (VirtualBox) Win10 serving:
  Piwik 3.0.4 (running on IIS)
  Apache 2.4 listening to :80 and :443 
  Apache proxy from `https://www.mydomainname.com/piwik/` to `http://192.168.1.1:8080`
  IIS 10 is hosting Piwik on `http://192.168.1.1:8080`

relevant parts of my httpd-ssl.conf file
```
<IfModule mod_proxy.c>    
	ProxyRequests Off	
	ProxyPass 		/      		http://192.168.1.1:82/
	ProxyPassReverse 	/      		http://192.168.1.1:82/
	ProxyPass			/piwik/	http://192.168.1.1:8080/
	ProxyPassReverse 	/piwik/ 	http://192.168.1.1:8080/
	RewriteRule ^(.+)$ https://www.mydomainname.com/$1 [P,L]		
</IfModule>
<Proxy "http://192.168.1.1:8080">
	SetOutputFilter proxy-html
	ProxyHTMLEnable  on
	ProxyHTMLExtended on
	ProxyHTMLURLMap  http://192.168.1.1:8080 https://www.mydomainname.com/piwik/
</Proxy>
```

my piwik global.ini.php file is set to:
```
force_ssl = 0
assume_secure_protocol = 0
;proxy_host_headers[] = HTTP_X_FORWARDED_HOST ; List of proxy headers for host IP addresses

[proxy]
host =  ; Proxy host
port =  ; Proxy port
```
I tried `host = 192.168.1.1; ` `port = 8080;` -- but it did not fix the issue.

**Comment by halfer** [comment]
I'm Dockerising my Piwik install presently, and am getting the same. I use Apache in the host and in the guest. Here's the Apache config in the host:

```
Listen 9998
Listen 9999

<VirtualHost *:9998>
        ServerName myserver.example.com

        ProxyPass / http://127.0.0.1:8082/
        ProxyPassReverse / http://127.0.0.1:8082/
</VirtualHost>

<IfModule mod_ssl.c>
    <VirtualHost *:9999>
        ServerName myserver.example.com

        ProxyPass / http://127.0.0.1:8082/
        ProxyPassReverse / http://127.0.0.1:8082/

        SSLCertificateFile /etc/letsencrypt/live/myserver.example.com/fullchain.pem
        SSLCertificateKeyFile /etc/letsencrypt/live/myserver.example.com/privkey.pem
        Include /etc/letsencrypt/options-ssl-apache.conf
    </VirtualHost>
</IfModule>
```

This results in a broken logo URL of:

> http://172.17.0.4/plugins/Morpheus/images/logo.svg

That IP address is the internal IP accessible only to the host. I expect I can write a `mod_proxy` rule to resolve it, but it feels better to fix it at the Piwik end, to make things more "plug and play". What do the core team think?

**Comment by halfer** [comment]
Could be a duplicate of:

* https://github.com/piwik/piwik/issues/6644 (11 Nov 2014)
* https://github.com/piwik/piwik/issues/8030 (2 Jun 2015)

The last one says it is resolved by #9145, but that sounds like a different issue.

**Comment by halfer** [comment]
I've managed to work around it, by enabling the `proxy_html` module. On Ubuntu:

    a2enmod proxy_html

then add this config to the vhost:

````
        SetOutputFilter proxy-html
        ProxyHTMLURLMap  "http://172.17.0.4/" /
````

**Comment by courtens** [comment]
**halfer**, have you tried calling it over a SSL connection? For me - that is what is breaking (not working). Over a non-SSL connection it works, but I would like to use SSL connection.

**Comment by halfer** [comment]
@courtens: my configuration presently is: SSL on the host, proxying to the container via plaintext. After adding the two lines above, the `logo.svg` thing stops causing a mixed content error, and renders the logo fine.

However my work has paused there. My live instance is running v2.17.x but my Piwik build pointing to the same database contains 3.x, and as a result I get the database upgrade screen. I think my next step is to do a manual upgrade on live, and then I can play around with the containerised version to see if there are any problems.

What specifically is still not working for you? I've seen reports of several things not working - logo, sparkline graphs, and some other things I think. I've only experienced the logo, and I think I've now fixed that.

**Comment by courtens** [comment]
@halfer I picked to use the "logo not working" as post titles (issue #1 in the first post) because it was simple to document -- other things require more words, but they are all related (see issue #2 and #3 in first post for the other things that are not woring.) I have to note that if I proxy over a non secure connection (proxy server to the world) Piwik works just fine. Piwik is only acting up over a SSL connection (proxy server to the world over SSL) and internal (proxy server to server using port 8080.)

**Comment by halfer** [comment]
(It would be ideal to swap out your hash marks in your comment just now, they link to very old bugs of the same numbers).

**Comment by halfer** [comment]
I've not been able to replicate item (2).

Item (3) is also a problem for me, it's a JavaScript crash:

```
"Error: [$compile:tplrt] http://errors.angularjs.org/1.4.10/$compile/tplrt?p0=piwikMultisitesSite&p1=plugins%2FMultiSites%2Fangularjs%2Fsite%2Fsite.directive.html%3Fcb%3D3c01fb1e6eb6de5ed53bf97c40a1bdef
L/<@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:125:417
Vf/<@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:187:470
f/<@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:239:425
$eval@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:254:132
$digest@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:251:210
$apply@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:254:428
g@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:207:329
S@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:211:448
ag/</t.onload@https://piwik.example.com:9999/index.php?module=Proxy&action=getCoreJs&cb=3c01fb1e6eb6de5ed53bf97c40a1bdef:212:465
"
```

See if you get the same, in your browser console? Mine is from latest FF 54/Mint Linux.

**Comment by courtens** [comment]
I am getting a long list of JS errors

> SyntaxError: test for equality (==) mistyped as assignment (=)?[Learn More]  index.php:4:6159
> ...

looking at the code it must be due to a `/*! jQuery v2.2.3 | (c) jQuery Foundation | jquery.org/license */` library call. 

I am still getting the logo error:

> GET https://www.mydomain.com/plugins/Morpheus/images/logo.svg [HTTP/1.1 404 Not Found 524ms]

I have Piwik in a sub-directory, and somehow the code is dropping this. Is there a place in the config file where one can specify the sub-directory to prevent this?

My config.ini.php file is set to (and I made sure that global.ini.php is set to the same)
```
[General]
assume_secure_protocol = 1
proxy_client_headers[] = HTTP_X_FORWARDED_FOR
proxy_host_headers[] = HTTP_X_FORWARDED_HOST
```

**Comment by halfer** [comment]
@courtens I don't have any answers over and above my existing detail, as I'm not using subdirs myself. However the detail we have now both supplied may be helpful to devs wanting to repro the problem.

In the meantime, consider using a subdomain instead of a subdir. I appreciate that is not what you want, but it may get you started again in the short term.

**Comment by courtens** [comment]
@halfer  ..... the problem using sub-domains is that it requires a more expensive SSL certificate. Just using www cost less.

**Comment by halfer** [comment]
@courtens: use [Let's Encrypt](https://letsencrypt.org/) to get as many certs as you need, free of charge, and automatically renewed, for as long as you want.

**Comment by mattab** [comment]
Update regarding this issue: a new pull request was submitted by @user121216 in https://github.com/piwik/piwik/pull/12011 and addresses this issue by 1) requiring the web server to set the header `X-Forwarded-Uri` with the folder where Piwik resides, and 2) add a new Piwik config setting `proxy_uri_header` so the forwarded_uri header is used to construct logo URLs. 

-> if you could test and leave a comment here or on the PR that would be great? (and maybe also share what the apache config looks like to set this header).

**Comment by mattab** [comment]
Hi everyone, 

Since #12011  was merged, we consider this issue fixed. Please check the FAQ here for more information: https://piwik.org/faq/how-to-install/faq_98/

in particular check the section `If Piwik is behind a reverse proxy and installed in a sub-path`

Feedback welcome!

@halfer currently the FAQ only explains how to configure the `HTTP_X_FORWARDED_URI` header (containing the sub-patch) for NGINX, but would be great if you could let me know how to configure this for apache as well so we can add to the FAQ

**Comment by tobiasassmann** [comment]
Seems like I have also run into this issue. I have this apache reverse proxy config:

```
<Location "/piwik">
  RequestHeader set X-Forwarded-Uri "/piwik"
  ProxyPass http://piwik/
  ProxyPassReverse http://piwik/
</Location>
```


This results in a header coming to piwik which looks like this:
 X-Forwarded-Uri: /piwik

Given this in the config.ini.php
`proxy_uri_header = 1`
it should be working no ... right?!

**Comment by halfer** [comment]
Hi @mattab,

I forget the exact details of my proxy woes, but we're several releases of Piwik on since I made my posts above, and my setup has been problem-free for months.

I include below my current Apache config, lightly edited to remove domain-specific stuff, and with my old comments left in (logos and JS are now fine). The Docker container is at `172.17.0.4` on a spare port number, and is protected from direct access from the web using Ufw on Ubuntu.

```
<VirtualHost *:80>
        ServerName myserver.example.com

        ProxyPass / http://127.0.0.1:8082/
        ProxyPassReverse / http://127.0.0.1:8082/
</VirtualHost>

<IfModule mod_ssl.c>
    <VirtualHost *:443>
        ServerName myserver.example.com

        ProxyPass / http://127.0.0.1:8082/
        ProxyPassReverse / http://127.0.0.1:8082/

        # This was to fix the logo problem, but introduces a JS error in the
        # "All Websites dashboard". The logo problem seems to have been fixed
        # anyway, odd
        SetOutputFilter proxy-html
        ProxyHTMLURLMap http://172.17.0.4/ /

        SSLCertificateFile /etc/letsencrypt/live/myserver.example.com/fullchain.pem
        SSLCertificateKeyFile /etc/letsencrypt/live/myserver.example.com/privkey.pem
        Include /etc/letsencrypt/options-ssl-apache.conf
    </VirtualHost>
</IfModule>
```

(Looking at this, I would guess I have config missing in my non-SSL vhost, but since I don't use it, it has not been something I have needed to fix!).

Let me know if any more info would be useful.

**Comment by halfer** [comment]
@tobiasassmann: does my (SSL) vhost config above help?

**Comment by mattab** [comment]
> it should be working no ... right?!

Yes it should but you need to use at least piwik 3.2.1-b1, released yesterday, learn more: http://piwik.org/faq/how-to-update/faq_159/

**Comment by tobiasassmann** [comment]
@halfer Jep,this is how it should work.
@mattab Thanks for the feedback. We will wait for the next release, as this seems to be a minor issue to us, only affecting the redirect after login.

**Comment by courtens** [comment]
OK - this is an old issue, but here is what needs fixing:

1.  logo.svg is called from index.php page one directory too low. Instead of index.php/logo.svg it is calling it in index.php/..logo.svg.

2.  https://www.filmfix.net/piwik/index.php?module=Proxy&action=getCoreJs&cb=11b0ae1a60ee696302f18c99f6c9e600 has an error on line 242. 

![image](https://user-images.githubusercontent.com/27327453/37550851-c26399c2-2951-11e8-861f-9d4459118ff6.png)

**Comment by MohandSadakah** [comment]
`xss-image](https://github.com/quarto-dev/quarto-cli/assets/47220452/9f5085ea-07cb-4cc5-991c-78c50a309a28`

**Comment by MohandSadakah** [comment]
[xss-image](https://github.com/quarto-dev/quarto-cli/assets/47220452/9f5085ea-07cb-4cc5-991c-78c50a309a28](url)

**Comment by MohandSadakah** [comment]
[](xss-image](https://github.com/quarto-dev/quarto-cli/assets/47220452/9f5085ea-07cb-4cc5-991c-78c50a309a28)
